bluetooth/src/commonMain/kotlin/device/DeviceState.kt

```diff
@@ -0,0 +1,344 @@
+                return { Idle(services, deviceInfo, connectionManager) }
+            }
+
+            override suspend fun afterOldStateIsRemoved(oldState: DeviceState) {
+                connectionManager.discoverServices()
+            }
+        }
+
+        data class Idle internal constructor(
+            val services: List<Service>,
+            override val deviceInfo: DeviceInfoImpl,
+            override val connectionManager: BaseDeviceConnectionManager
+        ) : Connected(deviceInfo, connectionManager) {
+
+            fun handleAction(action: DeviceAction): suspend () -> HandlingAction {
+                return { HandlingAction(action, emptyList(), services, deviceInfo, connectionManager) }
+            }
+        }
+
+        data class HandlingAction internal constructor(
+            internal val action: DeviceAction,
+            internal val nextActions: List<DeviceAction>,
+            val services: List<Service>,
+            override val deviceInfo: DeviceInfoImpl,
+            override val connectionManager: BaseDeviceConnectionManager
+        ) : Connected(deviceInfo, connectionManager), HandleAfterOldStateIsRemoved<DeviceState> {
+
+            fun addAction(newAction: DeviceAction): suspend () -> HandlingAction {
+                return { HandlingAction(action, listOf(*nextActions.toTypedArray(), newAction), services, deviceInfo, connectionManager) }
+            }
+
+            internal val actionCompleted = suspend {
+                when (action) {
+                    is DeviceAction.Read.Characteristic -> action.characteristic.updateValue()
+                    is DeviceAction.Read.Descriptor -> action.descriptor.updateValue()
+                    is DeviceAction.Write.Characteristic -> action.characteristic.updateValue()
+                    is DeviceAction.Write.Descriptor -> action.descriptor.updateValue()
+                    is DeviceAction.Notification -> {
+                        if (action.enable)
+                            action.characteristic.updateValue()
+                    }
+                }
+
+                if (nextActions.isEmpty()) {
+                    Idle(services, deviceInfo, connectionManager)
+                } else {
+                    val nextAction = nextActions.first()
+                    val remainingActions = nextActions.drop(1)
+                    HandlingAction(nextAction, remainingActions, services, deviceInfo, connectionManager)
+                }
+            }
+
+            override suspend fun afterOldStateIsRemoved(oldState: DeviceState) {
+                when (oldState) {
+                    is HandlingAction -> {
+                        if (oldState.action == action)
+                            return
+                    }
+                    else -> {}
+                }
+
+                connectionManager.performAction(action)
+            }
+        }
+
+        fun startDisconnected() {
```

Comment:
I'd name it `startDisconnecting` (like startConnecting)

Reply:
We can rethink some naming in the next pass